README.md
# admob_banner_stabilizer

A package to use banner ads in firebase_admob package easily.

![admob_widget_record](https://user-images.githubusercontent.com/12369062/102615893-058eeb00-417a-11eb-8f39-121d3ba865e0.gif)

## Features

### Banner ads appear in the same position as the Widget

The build method of AdmobBannerWidget simply returns a Container, but an ad with the same height is placed exactly on the Container.


### Widget height and ad height match, so it's easy to layout

This ensures that the ad and the button do not overlap.


### When the widget is removed from the widget tree, the ad disappears

The ad itself can be treated as if it were in the widget. You don't have to write an additional process to remove the ad.


### Navigator is supported

When another page is pushed from above, the ad disappears.


### Supports changing the orientation of the device

If you change the orientation of the device, the ad will be erased first, and then reappear in the new position.


### No crashes after consecutive screen transitions

You can dispose this widget continuously.


## Getting Started

This package depends on the firebase_admob plugin.

Please read the linked pages carefully and complete the necessary preparations.

[Add Firebase to your Flutter app](https://firebase.google.com/docs/flutter/setup?hl=en)
[firebase_admob plugin](https://pub.dev/packages/firebase_admob)

If you just want to run examples, these preparations is not necessary.


## Usage

Initialize the firebase_admob plugin before use AdMobBannerWidget.

```dart
void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await FirebaseAdMob.instance.initialize(
      appId: Platform.isAndroid
          // AdMob App ID for Firebase Demo Project (Flood-It!)
          ? 'ca-app-pub-8123415297019784~2664022459'
          // test GADApplicationIdentifier. see https://developers.google.com/admob/ios/quick-start?hl=en
          : 'ca-app-pub-3940256099942544~1458002511');
  runApp(MyApp());
}
```

Place the AdMobBannerWidget where you want the ad to appear.

```dart
class MyApp extends StatelessWidget {
  const MyApp();
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(),
        body: AdMobBannerWidget(
          adUnitId: BannerAd.testAdUnitId,
        ),
      ),
    );
  }
}
```

## Notes

### You cannot specify the size of the ads

The ad size will be the largest one that can be displayed within 1/8th of the height of the SafeArea.
AdMobBanner's SmartBanner standard is used as a reference.


### Ads cannot be moved as the widget is moved

They just appear and disappear.

You can't use it as a part of the list to scroll.


### You can't leave an ad running behind the scenes

It only calculates the position of the widget and puts the ad in front of it. You can't hide ads.

CHANGELOG.md
## [0.3.1] - 2021-01-26

* When a Widget, such as Drawer, temporarily covers an ad, the ad disappears.

## [0.3.0] - 2021-01-18

* Two new static methods, getAdSize and getBannerHeight, have been added.

## [0.2.0] - 2021-01-16

* It is now possible to specify the maxHeight of the AdMobBannerWidget.

## [0.1.0] - 2021-01-16

* Update a dependency to the latest release.

## [0.0.1] - 2020-12-20

* initial release.

test/unit_test.dart
import 'package:admob_banner_stabilizer/admob_banner_stabilizer.dart';
import 'package:firebase_admob/firebase_admob.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/mockito.dart';

void main() {
  SingleBanner singleBanner;
  setUp(() {
    // Note that SingleBanner is a singleton, so the same instance will be used for all test cases.
    singleBanner = SingleBanner(bannerAdConstructor: _mockBannerAdConstructor);
    singleBanner.show(
        callerHashCode: 12345678,
        adUnitId: null,
        size: null,
        anchorOffset: null,
        isMounted: null);
  });
  test('The last instance that generated the ad has ownership of it', () {
    expect(singleBanner.ownerHashCode, 12345678);
  });
  test('new bannerAd is set after calling show method', () {
    final old = singleBanner.bannerAd;
    expect(singleBanner.bannerAd == old, isTrue);
    singleBanner.show(
        callerHashCode: null,
        adUnitId: null,
        size: null,
        anchorOffset: null,
        isMounted: null);
    expect(singleBanner.bannerAd == old, isFalse);
  });
  test('bannerAd is null after SingleBanner is disposed', () {
    expect(singleBanner.bannerAd, isNotNull);
    singleBanner.dispose(callerHashCode: 12345678);
    expect(singleBanner.bannerAd, isNull);
  });
  test('Non-ownership instances cannot dispose of ads', () {
    expect(singleBanner.bannerAd, isNotNull);
    singleBanner.dispose(callerHashCode: 9999);
    expect(singleBanner.bannerAd, isNotNull);
  });
  test(
      'If another instance generates the ad, the previous owner loses ownership',
      () {
    expect(singleBanner.ownerHashCode, 12345678);
    singleBanner.show(
        callerHashCode: 8888,
        adUnitId: null,
        size: null,
        anchorOffset: null,
        isMounted: null);
    expect(singleBanner.ownerHashCode, 8888);
  });
}

class MockBannerAd extends Mock implements BannerAd {}

MockBannerAd _mockBannerAdConstructor({
  @required String adUnitId,
  @required AdSize size,
  MobileAdTargetingInfo targetingInfo,
  MobileAdListener listener,
}) {
  return MockBannerAd();
}


test/widget_test.dart
import 'package:firebase_admob/firebase_admob.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:admob_banner_stabilizer/admob_banner_stabilizer.dart';

void main() {
  testWidgets('works without Navigator', (WidgetTester tester) async {
    final adMobBannerWidget = AdMobBannerWidget(adUnitId: null);
    await tester.pumpWidget(MaterialApp(home: adMobBannerWidget));
    expect(find.byWidget(adMobBannerWidget), findsOneWidget);
  });

  testWidgets('works with Navigator', (WidgetTester tester) async {
    final adMobBannerWidget = AdMobBannerWidget(adUnitId: null);
    await tester.pumpWidget(MaterialApp(
      home: Navigator(
        observers: [RouteObserver()],
        pages: [MaterialPage(child: adMobBannerWidget)],
        onPopPage: (route, result) => route.didPop(result),
      ),
    ));
    expect(find.byWidget(adMobBannerWidget), findsOneWidget);
  });

  // This test case does not work. Please tell me how to make it work.
  testWidgets('throws if using with Navigator without RouteObserver',
      (WidgetTester tester) async {
    final adMobBannerWidget = AdMobBannerWidget(adUnitId: null);
    await tester.pumpWidget(MaterialApp(
      home: Navigator(
        // no observers here
        pages: [MaterialPage(child: adMobBannerWidget)],
        onPopPage: (route, result) => route.didPop(result),
      ),
    ));
    // await tester.pumpAndSettle();
    expect(tester.takeException(), isInstanceOf<StateError>());
  }, skip: true);

  testWidgets('about getAdSize and getBannerHeight static methods',
      (WidgetTester tester) async {
    AdSize _adSize;
    double _bannerHeight;
    Widget myWidget(Size size, double maxHeight) {
      return MaterialApp(
          home: MediaQuery(
        data: MediaQueryData(size: size),
        child: Builder(
          builder: (context) {
            _adSize = AdMobBannerWidget.getAdSize(
                context: context, maxHeight: maxHeight);
            _bannerHeight = AdMobBannerWidget.getBannerHeight(
                context: context, maxHeight: maxHeight);
            return Container();
          },
        ),
      ));
    }

    await tester.pumpWidget(myWidget(Size(25, 1000), null));
    expect(_bannerHeight, equals(50));
    expect(_adSize, equals(AdSize.banner));
    await tester.pumpWidget(myWidget(Size(728, 720), null));
    expect(_bannerHeight, equals(90));
    expect(_adSize, equals(AdSize.leaderboard));
    await tester.pumpWidget(myWidget(Size(468, 480), null));
    expect(_bannerHeight, equals(60));
    expect(_adSize, equals(AdSize.fullBanner));
    await tester.pumpWidget(myWidget(Size(320, 800), null));
    expect(_bannerHeight, equals(100));
    expect(_adSize, equals(AdSize.largeBanner));
    await tester.pumpWidget(myWidget(Size(1000, 1000), 50));
    expect(_bannerHeight, equals(50));
    expect(_adSize, equals(AdSize.banner));
    await tester.pumpWidget(myWidget(Size(1000, 1000), 49));
    expect(tester.takeException(), isInstanceOf<StateError>());
  });
}


test/admob_banner_stabilizer_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:admob_banner_stabilizer/admob_banner_stabilizer.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}

// テストする項目
// _SingleBanner
// // BannerAdをdisposeしたら、必ず、次をセットするかnullにする。
// 最後に広告を生成したインスタンスが所有権を持ち、そこからしかdisposeできない。
// 別のインスタンスが新たに広告生成を行った場合、所有権を失う。
// Navigator(またはRouteObserver)があっても正常に動く
// 同なくても正常に動く
// RouteObserverを与えずにNavigatorと一緒に使った場合に、わかりやすいエラーメッセージを出す


example/lib/main.dart
import 'dart:io';

import 'package:flutter/material.dart';

import 'package:firebase_admob/firebase_admob.dart';
import 'package:admob_banner_stabilizer/admob_banner_stabilizer.dart';

// This example is without Navigator.
void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await FirebaseAdMob.instance.initialize(
      appId: Platform.isAndroid
          // AdMob App ID for Firebase Demo Project (Flood-It!)
          ? 'ca-app-pub-8123415297019784~2664022459'
          // test GADApplicationIdentifier. see https://developers.google.com/admob/ios/quick-start?hl=en
          : 'ca-app-pub-3940256099942544~1458002511');
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp();
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyBody(),
    );
  }
}

class MyBody extends StatefulWidget {
  const MyBody({
    Key key,
  }) : super(key: key);

  @override
  _MyBodyState createState() => _MyBodyState();
}

class _MyBodyState extends State<MyBody> {
  bool isUpperPosition = true;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Column(
        children: [
          SizedBox(height: isUpperPosition ? 50 : 150),
          Text("↓↓↓↓↓↓↓↓↓↓↓↓↓ AD HERE ↓↓↓↓↓↓↓↓↓↓↓↓↓"),
          AdMobBannerWidget(
            adUnitId: BannerAd.testAdUnitId,
            backgroundColor: Colors.yellow,
          ),
          Text("↑↑↑↑↑↑↑↑↑↑↑↑↑ AD HERE ↑↑↑↑↑↑↑↑↑↑↑↑↑"),
          RaisedButton(
              child: Text('Change Ad Position'),
              onPressed: () {
                setState(() {
                  isUpperPosition = !isUpperPosition;
                });
              })
        ],
      ),
    );
  }
}


lib/single_banner.dart
import 'package:firebase_admob/firebase_admob.dart';
import 'package:flutter/foundation.dart';

/// A wrapper for BannerAd.
/// It could be called AdMobBannerRepository.
/// BannerAd triggers an error if you dispose it more than once, but there is no way to check if it has already been disposed.
/// This is a class to ensure that BannerAd is not disposed more than once.
/// Being a Singleton, it also ensures that there is only one instance of BannerAd at a time.
/// The instance that generated the ad is recorded, and no other instance can dispose the ad.
/// When a BannerAd is disposed of, it is immediately discarded and the next instance is set or nulled. This prevents BannerAd from being disposed of more than once.
class SingleBanner {
  /// Factory constructor
  factory SingleBanner({BannerAdConstructor bannerAdConstructor}) {
    _instance ??=
        SingleBanner._internal(bannerAdConstructor: bannerAdConstructor);
    return _instance;
  }

  /// Internal constructor
  SingleBanner._internal({BannerAdConstructor bannerAdConstructor})
      : _bannerAdConstructor =
            bannerAdConstructor ?? _bannerAdConstructorWrapper;

  /// The singleton instance
  static SingleBanner _instance;

  /// BannerAd constructor. it's overridden when testing.
  BannerAdConstructor _bannerAdConstructor;

  /// The BannerAd instance
  BannerAd _bannerAd;
  BannerAd get bannerAd => _bannerAd;

  /// This indicates which SingleBanner instance has the ownership of the BannerAd instance.
  int _ownerHashCode;
  int get ownerHashCode => _ownerHashCode;

  /// Shows the banner ad
  /// The instance that calls this method will get the ownership.
  void show({
    @required int callerHashCode,
    @required String adUnitId,
    @required AdSize size,
    @required double anchorOffset,
    @required bool isMounted,
  }) {
    // after disposing, _bannerAd must be null or the next instance must be set.
    _bannerAd?.dispose();
    _bannerAd = _bannerAdConstructor(
      adUnitId: adUnitId,
      size: size,
      listener: (MobileAdEvent event) {
        // Register a listener so that the show is only called after the load is complete.
        // Otherwise, if the screen changes between the time you call show and the time the load actually completes, you won't be able to delete the ad.
        if (event == MobileAdEvent.loaded) {
          if (isMounted) {
            _bannerAd.show(anchorOffset: anchorOffset);
          } else {
            _bannerAd = null;
          }
        }
      },
    );
    _ownerHashCode = callerHashCode;
    _bannerAd.load();
  }

  /// The last instance that generated the ad has the ownership, and the ad cannot be disposed of by other instances.
  void dispose({@required int callerHashCode}) {
    if (callerHashCode == _ownerHashCode) {
      // after disposing, _bannerAd must be null or the next instance must be set.
      _bannerAd?.dispose();
      _bannerAd = null;
    }
  }
}

/// This is the same as BannerAd constructor.
/// This function will be mocked during testing.
BannerAd _bannerAdConstructorWrapper({
  @required String adUnitId,
  @required AdSize size,
  MobileAdTargetingInfo targetingInfo,
  MobileAdListener listener,
}) {
  return BannerAd(
    adUnitId: adUnitId,
    size: size,
    targetingInfo: targetingInfo,
    listener: listener,
  );
}

/// Type representing the BannerAd constructor
typedef BannerAdConstructor = BannerAd Function({
  @required String adUnitId,
  @required AdSize size,
  MobileAdTargetingInfo targetingInfo,
  MobileAdListener listener,
});


example/lib/minimum.dart
import 'dart:io';

import 'package:flutter/material.dart';

import 'package:firebase_admob/firebase_admob.dart';
import 'package:admob_banner_stabilizer/admob_banner_stabilizer.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await FirebaseAdMob.instance.initialize(
      appId: Platform.isAndroid
          // AdMob App ID for Firebase Demo Project (Flood-It!)
          ? 'ca-app-pub-8123415297019784~2664022459'
          // test GADApplicationIdentifier. see https://developers.google.com/admob/ios/quick-start?hl=en
          : 'ca-app-pub-3940256099942544~1458002511');
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp();
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(),
        body: AdMobBannerWidget(
          adUnitId: BannerAd.testAdUnitId,
        ),
      ),
    );
  }
}


example/lib/with_fab.dart
import 'dart:io';

import 'package:flutter/material.dart';

import 'package:firebase_admob/firebase_admob.dart';
import 'package:admob_banner_stabilizer/admob_banner_stabilizer.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await FirebaseAdMob.instance.initialize(
      appId: Platform.isAndroid
          // AdMob App ID for Firebase Demo Project (Flood-It!)
          ? 'ca-app-pub-8123415297019784~2664022459'
          // test GADApplicationIdentifier. see https://developers.google.com/admob/ios/quick-start?hl=en
          : 'ca-app-pub-3940256099942544~1458002511');
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp();
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHome(),
    );
  }
}

class MyHome extends StatelessWidget {
  const MyHome({
    Key key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      floatingActionButton: Padding(
        padding: EdgeInsets.only(
            bottom: AdMobBannerWidget.getBannerHeight(context: context) + 24),
        child: FloatingActionButton(
          child: Icon(Icons.check_circle_rounded),
          onPressed: () {},
        ),
      ),
      body: SafeArea(
        child: Column(
          children: [
            Expanded(child: Container()),
            AdMobBannerWidget(
              adUnitId: BannerAd.testAdUnitId,
            ),
          ],
        ),
      ),
    );
  }
}


example/lib/specify_size.dart
import 'dart:io';

import 'package:flutter/material.dart';

import 'package:firebase_admob/firebase_admob.dart';
import 'package:admob_banner_stabilizer/admob_banner_stabilizer.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await FirebaseAdMob.instance.initialize(
      appId: Platform.isAndroid
          // AdMob App ID for Firebase Demo Project (Flood-It!)
          ? 'ca-app-pub-8123415297019784~2664022459'
          // test GADApplicationIdentifier. see https://developers.google.com/admob/ios/quick-start?hl=en
          : 'ca-app-pub-3940256099942544~1458002511');
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp();

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  double maxBannerHeight;
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(),
        body: Column(
          children: [
            SizedBox(height: 10),
            Row(
              children: [
                Text('Choose maxBannerHeight!'),
                Expanded(
                  child: SizedBox(
                    height: 40,
                    child: ListWheelScrollView(
                      itemExtent: 20,
                      children:
                          List.generate(15, (i) => Text((i * 10).toString())),
                      physics: FixedExtentScrollPhysics(),
                      onSelectedItemChanged: (value) {
                        setState(() {
                          maxBannerHeight = value.toDouble() * 10;
                        });
                      },
                    ),
                  ),
                ),
              ],
            ),
            SizedBox(height: 64),
            Container(
              height: maxBannerHeight,
              color: Colors.blue,
              child: AdMobBannerWidget(
                adUnitId: BannerAd.testAdUnitId,
                maxHeight: maxBannerHeight,
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/admob_banner_stabilizer.dart
/// A package to use banner ads in firebase_admob package easily.
library admob_banner_stabilizer;

import 'dart:async';

import 'package:firebase_admob/firebase_admob.dart';
import 'package:flutter/widgets.dart';

/// The main Widget for displaying banner ads. Place this Widget where you want to display the banner ad.
class AdMobBannerWidget extends StatefulWidget {
  // By forcing a rebuild with a UniqueKey, the ad will follow the changes in the widget state.
  // Note that if you prevent the rebuild by giving a different key, the ad may not follow the widget.
  AdMobBannerWidget({
    @required this.adUnitId,
    this.backgroundColor,
    this.maxHeight,
    Key key,
  }) : super(key: key ?? UniqueKey());

  /// Your ID of banner ad unit from AdMob.
  final String adUnitId;

  /// Background color of banner area. Default is transparent.
  final Color backgroundColor;

  /// Specifies the maximum value for the height of the banner.
  /// If not specified, it will be 1/8 of the height of the SafeArea.
  /// Note that this does not mean that banners with the specified height will be displayed.
  /// Values less than 50 will be treated as 50.
  final double maxHeight;

  /// Returns the AdSize that will be selected when the banner ad is displayed in the given context.
  /// In addition, you can also give the maxHeight.
  static AdSize getAdSize({@required BuildContext context, double maxHeight}) {
    final _viewPaddingTop = WidgetsBinding.instance.window.viewPadding.top /
        MediaQuery.of(context).devicePixelRatio;
    final _viewPaddingBottom =
        WidgetsBinding.instance.window.viewPadding.bottom /
            MediaQuery.of(context).devicePixelRatio;
    final _screenWidth = MediaQuery.of(context).size.width;
    if (maxHeight != null && maxHeight < 50) {
      throw StateError(
          'maxHeight of AdMobBannerWidget must be greater than 50 because there is no AdSize with a height less than 50.');
    }
    final _availableBannerHeight = maxHeight ??
        (MediaQuery.of(context).size.height -
                _viewPaddingTop -
                _viewPaddingBottom) /
            8;
    if (_screenWidth >= 728 && _availableBannerHeight >= 90) {
      return AdSize.leaderboard;
    } else if (_screenWidth >= 468 && _availableBannerHeight >= 60) {
      return AdSize.fullBanner;
    } else if (_screenWidth >= 320 && _availableBannerHeight >= 100) {
      return AdSize.largeBanner;
    } else {
      return AdSize.banner;
    }
  }

  /// Returns the AdSize height that will be selected when the banner ad is displayed in the given context.
  /// In addition, you can also give the maxHeight.
  static double getBannerHeight(
      {@required BuildContext context, double maxHeight}) {
    final _adSize = getAdSize(context: context, maxHeight: maxHeight);
    if (_adSize == AdSize.leaderboard) {
      return 90;
    } else if (_adSize == AdSize.fullBanner) {
      return 60;
    } else if (_adSize == AdSize.largeBanner) {
      return 100;
    } else if (_adSize == AdSize.banner) {
      return 50;
    } else {
      throw StateError('_adSize is strange.');
    }
  }

  @override
  _AdMobBannerWidgetState createState() => _AdMobBannerWidgetState();
}

class _AdMobBannerWidgetState extends State<AdMobBannerWidget> with RouteAware {
  /// The height of the banner ad.
  double _bannerHeight;

  /// The size of the banner ad. AdSize is defined in the firebase_admob plugin.
  AdSize _adSize;

  /// If this is true, this Widget shows the banner ad.
  bool _doShowAd = false;

  /// The timer to show banner ad. This timer is used to delay the display of ads in case there is a screen transition animation.
  Timer _timer;

  /// Loads and shows the banner ad.
  void _loadAndShowBanner() {
    assert(_bannerHeight != null);
    assert(_adSize != null);
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      // This timer is used to delay the display of ads in case there is a screen transition animation.
      _timer = Timer(Duration(seconds: 1), () {
        final RenderBox _renderBox = context.findRenderObject();
        final _isRendered = _renderBox.hasSize;
        if (_isRendered) {
          SingleBanner().show(
            isMounted: mounted,
            anchorOffset: _anchorOffset(),
            adUnitId: widget.adUnitId,
            callerHashCode: hashCode,
            size: _adSize,
          );
        }
      });
    });
  }

  /// This method is used to recalculate the position and size of the widget when the widget environment changes.
  void _refreshState() {
    _disposeBanner();
    _determineBannerSize();
    ModalRoute _route = ModalRoute.of(context);
    if (_route == null ||
        (_route.isCurrent && _route.willHandlePopInternally == false)) {
      _doShowAd = true;
    } else {
      _doShowAd = false;
    }
  }

  /// Chooses the largest ad within maxHeight or 1/8th of the height of the SafeArea, which is the area excluding notches and such.
  /// To clarify the height of the ad, SmartBanner is not used.
  void _determineBannerSize() {
    _adSize = AdMobBannerWidget.getAdSize(
        context: context, maxHeight: widget.maxHeight);
    _bannerHeight = AdMobBannerWidget.getBannerHeight(
        context: context, maxHeight: widget.maxHeight);
  }

  /// Calculates how many logical pixels above the bottom of the SafeArea this Widget is displayed.
  double _anchorOffset() {
    final RenderBox _renderBox = context.findRenderObject();
    assert(_renderBox.hasSize);
    final _y = _renderBox.localToGlobal(Offset.zero).dy;
    final _h = _renderBox.size.height;
    // Only the viewPadding was 0 when I used MediaQuery to get it, so I got it directly from the window.
    // Since it returns physical pixels, divide it by the devicePixelRatio to convert it to logical pixels.
    final _vpb = WidgetsBinding.instance.window.viewPadding.bottom /
        MediaQuery.of(context).devicePixelRatio;
    final _screenHeight = MediaQuery.of(context).size.height;
    return _screenHeight - _y - _h - _vpb;
  }

  @override
  Widget build(BuildContext context) {
    if (_doShowAd) {
      _loadAndShowBanner();
    }
    return Container(height: _bannerHeight, color: widget.backgroundColor);
  }

  @override
  void didChangeDependencies() {
    final _route = ModalRoute.of(context);
    if (_route != null &&
        // If _route is MaterialApp's default route, it doesn't throw.
        !(_route.settings.name == '/' && _route.settings.arguments == null) &&
        Navigator.of(context).widget.observers.isEmpty) {
      throw StateError(
          'Give an RouteObserver when using AdMobBannerWidget with Navigator.');
    }
    if (_route != null && Navigator.of(context).widget.observers.isNotEmpty) {
      // If there is more than one Observer, we may need to determine if first is okay and change the code
      _routeObserver = Navigator.of(context).widget.observers.first;
      _routeObserver.subscribe(this, _route);
    }
    _refreshState();
    super.didChangeDependencies();
  }

  @override
  void dispose() {
    _disposeBanner();
    _routeObserver?.unsubscribe(this);
    _timer.cancel();
    super.dispose();
  }

  void _disposeBanner() {
    _timer?.cancel();
    SingleBanner().dispose(callerHashCode: hashCode);
  }

  RouteObserver<dynamic> _routeObserver;

  @override
  void didPushNext() {
    setState(() {
      _refreshState();
    });
  }

  @override
  void didPopNext() {
    setState(() {
      _refreshState();
    });
  }

  @override
  void didPop() {
    setState(() {
      _refreshState();
    });
  }
}

/// A wrapper for BannerAd.
/// It could be called AdMobBannerRepository.
///
/// BannerAd triggers an error if you dispose it more than once, but there is no way to check if it has already been disposed.
///
/// This is a class to ensure that BannerAd is not disposed more than once.
/// Being a Singleton, it also ensures that there is only one instance of BannerAd at a time.
/// The instance that generated the ad is recorded, and no other instance can dispose the ad.
///
/// When a BannerAd is disposed of, it is immediately discarded and the next instance is set or nulled. This prevents BannerAd from being disposed of more than once.
class SingleBanner {
  /// Factory constructor
  factory SingleBanner({
    BannerAdConstructor bannerAdConstructor,
  }) {
    _instance ??=
        SingleBanner._internal(bannerAdConstructor: bannerAdConstructor);
    return _instance;
  }

  /// Internal constructor
  SingleBanner._internal({
    BannerAdConstructor bannerAdConstructor,
  }) : _bannerAdConstructor =
            bannerAdConstructor ?? _bannerAdConstructorWrapper;

  /// The singleton instance
  static SingleBanner _instance;

  /// BannerAd constructor. it's overridden when testing.
  BannerAdConstructor _bannerAdConstructor;

  BannerAd _bannerAd;

  /// The BannerAd instance
  BannerAd get bannerAd => _bannerAd;

  int _ownerHashCode;

  /// This indicates which SingleBanner instance has the ownership of the BannerAd instance.
  int get ownerHashCode => _ownerHashCode;

  /// Shows the banner ad
  /// The instance that calls this method will get the ownership.
  void show({
    @required int callerHashCode,
    @required String adUnitId,
    @required AdSize size,
    @required double anchorOffset,
    @required bool isMounted,
  }) {
    // after disposing, _bannerAd must be null or the next instance must be set.
    _bannerAd?.dispose();
    _bannerAd = _bannerAdConstructor(
      adUnitId: adUnitId,
      size: size,
      listener: (MobileAdEvent event) {
        // Register a listener so that the show is only called after the load is complete.
        // Otherwise, if the screen changes between the time you call show and the time the load actually completes, you won't be able to delete the ad.
        if (event == MobileAdEvent.loaded) {
          if (isMounted) {
            _bannerAd.show(anchorOffset: anchorOffset);
          } else {
            _bannerAd = null;
          }
        }
      },
    );
    _ownerHashCode = callerHashCode;
    _bannerAd.load();
  }

  /// The last instance that generated the ad has the ownership, and the ad cannot be disposed of by other instances.
  void dispose({@required int callerHashCode}) {
    if (callerHashCode == _ownerHashCode) {
      // after disposing, _bannerAd must be null or the next instance must be set.
      _bannerAd?.dispose();
      _bannerAd = null;
    }
  }
}

/// This is the same as BannerAd constructor.
/// This function will be mocked during testing.
BannerAd _bannerAdConstructorWrapper({
  @required String adUnitId,
  @required AdSize size,
  MobileAdTargetingInfo targetingInfo,
  MobileAdListener listener,
}) {
  return BannerAd(
    adUnitId: adUnitId,
    size: size,
    targetingInfo: targetingInfo,
    listener: listener,
  );
}

/// Type representing the BannerAd constructor
typedef BannerAdConstructor = BannerAd Function({
  @required String adUnitId,
  @required AdSize size,
  MobileAdTargetingInfo targetingInfo,
  MobileAdListener listener,
});


example/lib/specify_max_height.dart
import 'dart:io';

import 'package:flutter/material.dart';

import 'package:firebase_admob/firebase_admob.dart';
import 'package:admob_banner_stabilizer/admob_banner_stabilizer.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await FirebaseAdMob.instance.initialize(
      appId: Platform.isAndroid
          // AdMob App ID for Firebase Demo Project (Flood-It!)
          ? 'ca-app-pub-8123415297019784~2664022459'
          // test GADApplicationIdentifier. see https://developers.google.com/admob/ios/quick-start?hl=en
          : 'ca-app-pub-3940256099942544~1458002511');
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp();

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  double maxBannerHeight;
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(),
        body: Column(
          children: [
            SizedBox(height: 10),
            Row(
              children: [
                Text('Choose max banner height'),
                Expanded(
                  child: SizedBox(
                    height: 40,
                    child: ListWheelScrollView(
                      itemExtent: 20,
                      children:
                          List.generate(15, (i) => Text((i * 10).toString())),
                      physics: FixedExtentScrollPhysics(),
                      onSelectedItemChanged: (value) {
                        setState(() {
                          maxBannerHeight = value.toDouble() * 10;
                        });
                      },
                    ),
                  ),
                ),
              ],
            ),
            SizedBox(height: 64),
            Container(
              height: maxBannerHeight,
              color: Colors.blue,
              child: AdMobBannerWidget(
                adUnitId: BannerAd.testAdUnitId,
                maxHeight: maxBannerHeight,
              ),
            ),
          ],
        ),
      ),
    );
  }
}


example/lib/main2.dart
import 'dart:io';

import 'package:flutter/material.dart';

import 'package:firebase_admob/firebase_admob.dart';
import 'package:admob_banner_stabilizer/admob_banner_stabilizer.dart';
import 'package:provider/provider.dart';

// This example is with Navigator.
void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await FirebaseAdMob.instance.initialize(
      appId: Platform.isAndroid
          // AdMob App ID for Firebase Demo Project (Flood-It!)
          ? 'ca-app-pub-8123415297019784~2664022459'
          // test GADApplicationIdentifier. see https://developers.google.com/admob/ios/quick-start?hl=en
          : 'ca-app-pub-3940256099942544~1458002511');
  runApp(ChangeNotifierProvider(
    create: (BuildContext context) => PagesNotifier(),
    child: MyApp(),
  ));
}

class MyApp extends StatelessWidget {
  const MyApp();
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Navigator(
        observers: [RouteObserver<PageRoute<dynamic>>()],
        pages: [
          MaterialPage(key: ValueKey('MyBody'), child: MyBody()),
          if (Provider.of<PagesNotifier>(context).myBody2isPushed)
            MaterialPage(key: ValueKey('MyBody2'), child: MyBody2()),
        ],
        onPopPage: (route, result) {
          if (!route.didPop(result)) {
            return false;
          }
          return true;
        },
      ),
    );
  }
}

class MyBody extends StatefulWidget {
  const MyBody({
    Key key,
  }) : super(key: key);

  @override
  _MyBodyState createState() => _MyBodyState();
}

class _MyBodyState extends State<MyBody> {
  bool isUpperPosition = true;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('MyBody'),
      ),
      drawer: Drawer(
          child: ListView(children: [
        DrawerHeader(
          child: Text('Drawer'),
        ),
      ])),
      body: Column(
        children: [
          SizedBox(height: isUpperPosition ? 50 : 150),
          Text("↓↓↓↓↓↓↓↓↓↓↓↓↓ AD HERE ↓↓↓↓↓↓↓↓↓↓↓↓↓"),
          AdMobBannerWidget(
            adUnitId: BannerAd.testAdUnitId,
            backgroundColor: Colors.yellow,
          ),
          Text("↑↑↑↑↑↑↑↑↑↑↑↑↑ AD HERE ↑↑↑↑↑↑↑↑↑↑↑↑↑"),
          RaisedButton(
              child: Text('Change Ad Position'),
              onPressed: () {
                setState(() {
                  isUpperPosition = !isUpperPosition;
                });
              }),
          RaisedButton(
            child: Text('To MyBody2'),
            onPressed: () {
              Provider.of<PagesNotifier>(context, listen: false).pushMyBody2();
            },
          ),
          Builder(
            builder: (context) => RaisedButton(
              child: Text('Show the Drawer'),
              onPressed: () => Scaffold.of(context).openDrawer(),
            ),
          ),
        ],
      ),
    );
  }
}

class MyBody2 extends StatelessWidget {
  const MyBody2({
    Key key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('MyBody2'),
      ),
      body: Column(
        children: [
          SizedBox(height: 100),
          Text("↓↓↓↓↓↓↓↓↓↓↓↓↓ AD HERE ↓↓↓↓↓↓↓↓↓↓↓↓↓"),
          AdMobBannerWidget(
            adUnitId: BannerAd.testAdUnitId,
            backgroundColor: Colors.yellow,
          ),
          Text("↑↑↑↑↑↑↑↑↑↑↑↑↑ AD HERE ↑↑↑↑↑↑↑↑↑↑↑↑↑"),
          const SizedBox(height: 24),
          Text('tap the BACK button.'),
        ],
      ),
    );
  }
}

class PagesNotifier extends ChangeNotifier {
  bool myBody2isPushed = true;
  void pushMyBody2() {
    myBody2isPushed = true;
    notifyListeners();
  }
}
